guidelines/design-document-template.md

```diff
@@ -0,0 +1,35 @@
+# Design Document - My proposal 
+
+Author : **myemail@email.com**
+Date : Oct 2020
+
+## Overview 
+
+### Motivation
+Motivation to have this design document 
+
+### Examples &  Prior Art 
+Examples or references in how other solved similar or same problems
+
+```
+Hi,{mf}!
+```
+
+
+## Design 
+
+Discuss any background and motivation not already in the summary above. Give implementation details and all posible alternatives to solve the problem including affectations to other components.
+
```

Comment:
What needs to be in a design document:

# Goals and non-goals

# Alternatives considered and trade-offs

Since we got to a design document it means there is no one solution that everybody agrees that is perfect.
So there are obviously trade-offs.

This would be the "meat" of the design document, as it explains how the trade-offs are balanced.


## APIs  (probably sub-section for each alternative in "Alternatives considered" 

The APIs section should include signatures.
Should also include explanations on what they do, if not obvious.
Behavior described as pseudo-code.
But behavior is not implementation.

It is also useful to show how the APIs are invoked by the developer, the proper sequence in normal use.